### ISSUE TITLE
Suggest adding constructor inside class as well, not only when cursor is on class name

### ISSUE THREAD

**Issue by weitzhandler** [dotnet/roslyn#48318]
**Brief description:**

Inheriting from an abstract class that lacks a parameterless constructor results in a compile error, unless a constructor is added.
When hovering over the class name and hitting <kbd>Ctrl</kbd>+<kbd>.</kbd>, the suggestion "Generate constructor <constructor signature>" appears.

My suggestion is that this suggestion should also appear in the context menu for a <kbd>Ctrl</kbd>+<kbd>.</kbd> press anywhere in this class as long as a constructor is absent (same as when cursor is on class name).

**Languages applicable:**

Reproduced in C#, but might be applicable in VB too.

**Code example that the analyzer should report:**

![image](https://user-images.githubusercontent.com/2716316/95036742-8b4ddf00-06d1-11eb-9376-0cbdd25ea94e.png)
---
![image](https://user-images.githubusercontent.com/2716316/95036561-0367d500-06d1-11eb-882f-9c526c855751.png)
---
The constructor suggestion should also appear in the following screenshot (initiated anywhere in the class scope:
![image](https://user-images.githubusercontent.com/2716316/95036661-4aee6100-06d1-11eb-9357-74e524b1aa15.png)

**Comment by CyrusNajmabadi** [comment]
Yes.  this makes sense.  For anyone interested in doing this, we have helpers to ask/answer "is the caret between members in a type".  if so, then offering this refactoring would make sense.

**Comment by weitzhandler** [comment]
Gonna try that.

**Comment by weitzhandler** [comment]
I'm on it.

Does this test represent what we want (will add more cases, of course):

```c#
        [WorkItem(48318, "https://github.com/dotnet/roslyn/issues/48318")]
        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateDefaultConstructors)]
        public async Task TestGenerateConstructorFromProtectedConstructorCursorInClass()
        {
            await TestInRegularAndScriptAsync(
@"class C : B
{[||]
}

abstract class B
{
    protected B(int x)
    {
    }
}",
@"class C : B
{
    public C(int x) : base(x)
    {
    }
}

abstract class B
{
    protected B(int x)
    {
    }
}");
        }
```

**Comment by CyrusNajmabadi** [comment]
I would actually say that it should nto be offered here:

```
class C : B
{[||]
}
```

Since this is noton a blank line between members.   I would add a few more cases (i.e. have a type with some fields/methods in it, not just empty).

**Comment by weitzhandler** [comment]
> Since this is noton a blank line between members.

The compilation error is already here, even if there are no members. What if there aren't gonna be any members, just the constructor boiling down?

**Comment by weitzhandler** [comment]
> Since this is noton a blank line between members.

The compilation error (CS7036) is already there even if there are no members. What if there aren't gonna be any members, just the constructor boiling down?
My suggestion is, if there is a compilation error (i.e. the base class constructor is protected), it should be offered anywhere in the class.

**Comment by CyrusNajmabadi** [comment]
>  it should be offered anywhere in the class.

Definitely do *not* do this :)   Consider something like:

```
class X : Y
{
   // thousands of lines
   void SomeMethod()
   {
         // dozens of lines
         // deeply nested here
         if ($$
   }
}
```

The user's headspsace is not thinking about the class right now, and they are contextually miles away from this error/squiggle.  Just because that error has been reported *somewhere* does not mean the fix for it should be everywhere.

**Comment by weitzhandler** [comment]
Agree.
How about we only show it in the space between `Y { /*space*/ $$ /*anything*/`?

**Comment by weitzhandler** [comment]
So that:

```c#
        [WorkItem(48318, "https://github.com/dotnet/roslyn/issues/48318")]
        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsGenerateDefaultConstructors)]
        public async Task TestGenerateConstructorFromProtectedConstructorCursorInClass()
        {
            await TestInRegularAndScriptAsync(
@"class C : B
{
    void X() { }[||]
}

abstract class B
{
    protected B(int x)
    {
    }
}",
@"class C : B
{
    void X() { }
}

abstract class B
{
    protected B(int x)
    {
    }
}");
        }
```

**Comment by weitzhandler** [comment]
In short, I say we show the suggestion if the cursor is anywhere between the class name (current implementation) and the first method/property etc. in class / end of class if none.
Do you agree @CyrusNajmabadi?

**Comment by CyrusNajmabadi** [comment]
I don't really agree. I would just stick with what we do elsewhere. Allow it on the header or in between members.

**Comment by weitzhandler** [comment]
But in between members you're facing the same problem you mentioned before:

```c#
public class A
{
  public int Int { get; set; }
  // another zillion properties and methods
  [||] Really?
}
```

**Comment by CyrusNajmabadi** [comment]
In general though we do have that for several type/member-oriented features.  i.e. someone is in a location they want to add a constructor, so they bring up the lightbulb.  We have that feature (among others) there.

**Comment by weitzhandler** [comment]
So anywhere in between member / class openspace?

**Comment by CyrusNajmabadi** [comment]
yes.  on blank lines between members in a type.  I believe we have a helper for this (IsBetweenTypeMembers). :)

**Comment by weitzhandler** [comment]
I've added PR #48503, but this will only work between types, which while useful, doesn't actually resolve my original issue: show the suggestion **before** any members.

Can I implement that too?

**Comment by weitzhandler** [comment]
Consider continuing this discussion in the PR.
